### PR TITLE
Feature: Add luban engine

### PR DESCRIPTION
**Pull request by zjiefee**

### CODE REVIEW

File: src/server/services/task-manager/generateGcode.js [hunk 1]
```diff
@@ -135,14 +135,25 @@ export const generateGcode = (toolPaths, onProgress) => {
...
-                estimatedTime *= gcodeConfig.multiPasses;
+                estimatedTime = toolPathObj.estimatedTime * gcodeConfig.multiPasses;
+            } else {
+                estimatedTime += toolPathObj.estimatedTime;
             }
 
             isRotate = toolPathObj.isRotate;
             diameter = toolPathObj.diameter;
 
+            const { from, positionX = 0, positionY = 0, rotationB = 0 } = toolPathObj;
```
Comment (parachvte) [inline src/server/services/task-manager/generateGcode.js:147]: Where does `from` comes from?

File: src/app/flux/machine/index.js [hunk 2]
```diff
@@ -969,6 +971,10 @@ export const actions = {
     updateShouldCheckForUpdate: (shouldCheckForUpdate) => (dispatch) => {
         dispatch(baseActions.updateState({ shouldCheckForUpdate: shouldCheckForUpdate }));
         machineStore.set('shouldCheckForUpdate', shouldCheckForUpdate);
+    },
+    updateMultipleEngine: () => (dispatch, getState) => {
```
Comment (parachvte) [inline src/app/flux/machine/index.js:975]: weird naming, `updateMultipleEngine` -> `toggleMultipleEngine`?

File: src/server/services/task-manager/generateToolPath.js [hunk 3]
```diff
@@ -148,17 +147,49 @@ const generateCncToolPath = async (modelInfo, onProgress) => {
...
+const generateLaserToolPathFromEngine = async (modelInfo, onProgress) => {
+    const { type, sourceType } = modelInfo;
+    if ([TOOLPATH_TYPE_VECTOR + SOURCE_TYPE_RASTER].includes(type + sourceType)) {
+        const result = await editorProcess(modelInfo);
+        modelInfo.uploadName = result.filename;
+    }
+
+    modelInfo.gcodeConfig.stepOver = 1 / modelInfo.gcodeConfig.density;
+
+    modelInfo.toolpathFileName = pathWithRandomSuffix(`${path.parse(modelInfo.uploadName).name}.json`);
```
Comment (parachvte) [inline src/server/services/task-manager/generateToolPath.js:159]: We have many hard coded parameters. Where is the definition of Luban Engine API?

**Review by parachvte — CHANGES_REQUESTED** [review]
API needed

**Review by parachvte — APPROVED** [review]

**Review by meiliang1123 — APPROVED** [review]

**Pull request merged**